tensorboard/plugins/profile/pod_viewer/topology_graph/topology-graph.ts

```diff
@@ -0,0 +1,544 @@
+    runEnvironment_:
+        {type: Object, observer: 'runEnvironmentChanged_', value: null},
+    topoData_: {type: Object},
+    linkData_: {type: Object},
+    tpuType_: {type: String, observer: 'updateSystemInfo_'},
+    hostXStride_: {type: Number, value: 2},
+    hostYStride_: {type: Number, value: 2},
+    nodesPerChip_: {type: Number, value: 2},
+    hostGridWidth_: {type: Number},
+    hostGridHeight_: {type: Number},
+    chipGridSize_: {type: Number},
+    nodeGridHeight_: {type: Number},
+    nodeGridWidth_: {type: Number},
+    chipToChipMargin_: {type: Number},
+    hostToChipMargin_: {type: Number},
+    hostToHostMargin_: {type: Number},
+    xDimension_: {type: Number},
+    yDimension_: {type: Number},
+    totalCoreCount_ : {type: Number},
+    ready_: {type: Boolean, value: false},
+    metrics: {type: Array, notify: true, value: null},
+    allChannels: {type: Array, notify: true},
+    selectedChipId: {type: Number, notify: true},
+    selectedMetricIdx: {type: Number, value: 0},
+    selectedMetricLabel: {
+      type: String,
+      computed: 'getSelectedMetricLabel_(metrics, selectedMetricIdx)'
+    },
+    selectedChannelId:
+        {type: Number, value: 0, observer: 'selectedChannelIdChanged_'},
+    minChannelId: {type: Number, value: 0},
+    maxChannelId: {type: Number, value: 0},
+    gLink_: {type: Object},
+    selectedChannel: {type: Array, notify: true},
+    activeBarChartEle:
+        {type: Object, notify: true, observer: 'activeBarChartEleChanged_'}
+  },
+  observers:
+    ['updateAllData(data_, runEnvironment_, metrics, selectedMetricIdx)',
+     'updateTopology(topoData_, ready_)',
+     'updateLinks(linkData_, ready_)',
+    ],
+  /**
+   * Main function to draw topology graph based on TPU topology.
+   */
+  topologyGraph: function(data) {
+    d3.selectAll('#tpgraph g > *').remove();
+    d3.select('#tpgraph svg').remove();
+    d3.select('#tpgraph.svg-container').remove();
+    const margin = {top: 50, right: 0, bottom: 100, left: 30};
+    const width = 1620;
+    this.chipGridSize_ = 30;
+    this.chipToChipMargin_ = 10;
+    this.hostToChipMargin_ = 15;
+    this.hostToHostMargin_ = 10;
+    this.hostGridWidth_ = this.getHostGridSize(this.hostXStride_);
+    this.hostGridHeight_ = this.getHostGridSize(this.hostYStride_);
+    this.nodeGridWidth_ = this.chipGridSize_ / this.nodesPerChip_;
+    this.nodeGridHeight_ = this.chipGridSize_;
+    const hostXDim = this.xDimension_ / this.hostXStride_;
+    const hostYDim = this.yDimension_ / this.hostYStride_;
+    const colors = [
+      '#ffffd9', '#edf8b1', '#c7e9b4', '#7fcdbb', '#41b6c4', '#1d91c0',
+      '#225ea8', '#253494', '#081d58'
+    ];
+    const colorScale =
+        d3.scaleQuantile<string>().domain([0, 1.0]).range(colors);
+    const chipXDims = Array.from(Array(this.xDimension_).keys());
+    const chipYDims = Array.from(Array(this.yDimension_).keys());
+    let svg =
+        d3.select('#tpgraph')
```

Comment:
this is a global querySelector and should NOT be used.

Reply:
Not sure how to make this selector not global?

Reply:
You can use `this.$.tpgraph` to get a reference to the DOM node, and
then use `d3.select(this.$.tpgraph)` to convert that to a D3 selection.

See: <https://github.com/d3/d3-selection#select>

Reply:
Done, thanks @wchargin